pattern/variable/Variable.java

```diff
@@ -74,6 +81,31 @@ public ThingVariable asThing() {
         throw GraknException.of(INVALID_CASTING, className(this.getClass()), className(ThingVariable.class));
     }
 
+    public void addHints(Set<Label> labels) {
+        typeHints.addAll(labels);
+    }
+
+    public void retainHints(Set<Label> labels) {
+        typeHints.retainAll(labels);
+    }
+
+    public void removeHint(Label label) {
+        assert !(isSatisfiable && typeHints.isEmpty());
```

Comment:
i think it'll be more readable with two negated ORs, but this is subjective :D

Reply:
In general, I think `!(A ^ B)` and `(!A v !B)` are equally acceptable imho.